torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -13,6 +13,106 @@
 _NO_PARAM_UPDATE = None
 
 
+def _perform_local_step(
+    bucket: dist.GradBucket,
+    zero: ZeroRedundancyOptimizer,
+    rank: int,
+):
+    r"""
+    Performs a local optimizer step using the gradients provided by ``bucket``.
+
+    Arguments:
+        bucket (dist.GradBucket): the bucket providing the gradients.
+        zero (ZeroRedundancyOptimizer): the :class:`ZeroRedundancyOptimizer`
+            instance to perform the :meth:`_local_step`.
+        rank (int): the calling process's rank.
+
+    .. warning::
+        This function assumes that appropriate synchronization has taken place
+        so that the bucket's gradients can be used.
+    """
+    overlap_info = zero._overlap_info
+    bucket_index = bucket.index()
+    assert len(zero.optim.param_groups) == 1, \
+        "Overlapping DDP with ZeRO only supports a single parameter group"
+
+    # Construct the `gradients` input for the local optimizer step, which
+    # expects `None` in a list position to indicate that the corresponding
+    # parameter should not be updated
+    num_local_optim_params = len(zero.optim.param_groups[0]["params"])
+    gradients: List[Optional[torch.Tensor]] = \
+        [_NO_PARAM_UPDATE for _ in range(num_local_optim_params)]
+    assert bucket_index in overlap_info.offsets, \
+        f"Bucket index {bucket_index} was not assigned to rank {rank}"
+    offset = overlap_info.offsets[bucket_index]
+    bucket_gradients = bucket.gradients()
+    for i, grad in enumerate(bucket_gradients):
+        gradients[offset + i] = grad
+
+    zero._local_step(gradients)
+
+
+def _broadcast_bucket(
+    bucket_index: int,
+    zero: ZeroRedundancyOptimizer,
+    source_rank: int,
+):
+    r"""
+    Broadcasts a bucket's parameters.
+
+    Arguments:
+        bucket_index (int): the index of the bucket corresponding to the
+            parameters to broadcast.
+        zero (ZeroRedundancyOptimizer): the calling process's
+            :class:`ZeroRedundancyOptimizer` instance.
+        source_rank (int): the rank with the updated parameters, serving as the
+            source for the broadcast.
+    """
+    overlap_info = zero._overlap_info
+    device = overlap_info.params_per_bucket[bucket_index][0].device
+    device_index = zero._device_to_device_index[device]
+    assert bucket_index in zero._buckets[device_index][source_rank]
+    overlap_info.broadcast_handles.append(
+        dist.broadcast(
+            zero._buckets[device_index][source_rank][bucket_index],
+            src=source_rank,
+            async_op=True
+        )
+    )
+
+def _collect_ddp_bucket_info(
+    bucket: dist.GradBucket,
+    zero: ZeroRedundancyOptimizer,
+    rank: int,
+    rank_to_update: int,
```

Comment:
`owner_rank`?

And also, would I be correct if I assume `owner_rank` is always the same as the `source_rank` argument in `_broadcast_bucket`, as the owner of those params should both update and broadcast. If yes, let's consolidate these two args to use the same name.